### PR TITLE
New tutorial - Transparent Proxy and Filtering on K8S with Initializers

### PR DESCRIPTION
**Pull request by danisla**
Similar to the previous tutorial in #251 but this guide focuses on how to use the new Deployment Initializers to automate the configuration.

- [x] Tech writing / styleguide review - @ccarpentiere
- [x] Subject matter review - @ahmetb
- [x] Formatting review for publishing - @jmdobry
- [x] #251 has been merged to `master`

### CODE REVIEW

File: tutorials/transparent-proxy-and-filtering-on-k8s-with-initializers/index.md [hunk 1]
```diff
@@ -0,0 +1,216 @@
...
+title: Transparent Proxy and Filtering on Kubernetes with Initializers
+description: Learn how to run a transparent proxy on Kubernetes to filter and intercept traffic out of your deployments using custom deployment initializers.
+author: danisla
+tags: Kubernetes mitmproxy proxy
+date_published: 2017-09-23
+---
+
+Dan Isla | Google Cloud Solution Architect | Google
+
+This is a follow-on tutorial to the [Transparent Proxy and Filtering on K8S](https://cloud.google.com/community/tutorials/transparent-proxy-and-filtering-on-k8s) tutorial and shows how to simplify the application of transparent proxy for existing deployments using a Deployment Initializer. Initializers are one of the [Dynamic Admission Control](https://kubernetes.io/docs/admin/extensible-admission-controllers/) features of Kubernetes and available as an alpha feature as of Kubernetes 1.7.
```
Comment (ccarpentiere) [inline tutorials/transparent-proxy-and-filtering-on-k8s-with-initializers/index.md]: This is a follow-on tutorial to the [Transparent Proxy and Filtering on K8S](https://cloud.google.com/community/tutorials/transparent-proxy-and-filtering-on-k8s) tutorial. It shows how to simplify the application of a transparent proxy for existing deployments using a Deployment Initializer. Initializers are one of the [Dynamic Admission Control](https://kubernetes.io/docs/admin/extensible-admission-controllers/) features of Kubernetes, and are available as an alpha feature in Kubernetes 1.7.

File: tutorials/transparent-proxy-and-filtering-on-k8s-with-initializers/index.md [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+
+This tutorial uses the [tproxy-initializer](https://github.com/danisla/kubernetes-tproxy/tree/master/cmd/tproxy-initializer) Kubernetes Initializer to inject the sidecar InitContainer, ConfigMap and environment variables into a deployment when the annotation `"initializer.kubernetes.io/tproxy": "true"` is present. This tutorial also demonstrates how to deploy the tproxy Helm chart with the optional [Role Based Access Control](https://kubernetes.io/docs/admin/authorization/rbac/) support.
```
Comment (ccarpentiere) [inline tutorials/transparent-proxy-and-filtering-on-k8s-with-initializers/index.md]: This tutorial also demonstrates how to deploy the tproxy Helm chart with optional [Role Based Access Control](https://kubernetes.io/docs/admin/authorization/rbac/) support.

Comment (ccarpentiere) [inline tutorials/transparent-proxy-and-filtering-on-k8s-with-initializers/index.md]: Maybe add a link to tproxy Helm chart, unless your audience will typically already know what that is.

File: tutorials/transparent-proxy-and-filtering-on-k8s-with-initializers/index.md [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+
+Just like in the previous tutorial, the purpose of the [tproxy-sidecar](https://github.com/danisla/kubernetes-tproxy/tree/master/sidecar) container is to create firewall rules in the pod network to block egress traffic. The [tproxy-podwatch](https://github.com/danisla/kubernetes-tproxy/tree/master/cmd/tproxy-podwatch) controller watches for pod changes containing the annotation and automatically add/removes the local firewall `REDIRECT` rules to apply the transparent proxy to the pod.
+
+![architecture diagram](./tproxy_initializers_diagram.png)
+
+**Figure 1.** transparent proxy with initializers architecture diagram
+
+This tutorial also demonstrates how to use
```
Comment (ccarpentiere) [inline tutorials/transparent-proxy-and-filtering-on-k8s-with-initializers/index.md]: Complete or remove

File: tutorials/transparent-proxy-and-filtering-on-k8s-with-initializers/index.md [hunk 4]
```diff
@@ -0,0 +1,216 @@
...
+
+**Figure 1.** transparent proxy with initializers architecture diagram
+
+This tutorial also demonstrates how to use 
+
+## Objectives
+
+- Create a Kubernetes cluster with initializer and RBAC support using Google Container Engine
+- Deploy the tproxy, tproxy-initializer and the tproxy-podwatch pods using Helm
+- Deploy example apps with annotations to test external access to a Cloud Storage bucket
```
Comment (ccarpentiere) [inline tutorials/transparent-proxy-and-filtering-on-k8s-with-initializers/index.md]: Should be "Google Cloud Storage" on first reference, fine to use "Cloud Storage" for subsequent references.

File: tutorials/transparent-proxy-and-filtering-on-k8s-with-initializers/index.md [hunk 5]
```diff
@@ -0,0 +1,216 @@
...
+
+## Before you begin
+
+This tutorial assumes you already have a GCP account and are familiar with the high level concepts of Kubernetes Pods and Deployments.
```
Comment (ccarpentiere) [inline tutorials/transparent-proxy-and-filtering-on-k8s-with-initializers/index.md]: Should be "Google Cloud Platform (GCP)" on first reference, fine to use just "GCP" on subsequent  references.

File: tutorials/transparent-proxy-and-filtering-on-k8s-with-initializers/index.md [hunk 6]
```diff
@@ -0,0 +1,216 @@
...
+2. Clone the repository containing the code for this tutorial:
+
+        git clone https://github.com/danisla/kubernetes-tproxy
+        cd kubernetes-tproxy
+
+    The remainder of this tutorial will be run from the root of the cloned repository directory.
+
+## Create Container Engine cluster and install Helm
+
+1. Create Container Engine cluster with alpha features enabled, RBAC support and a cluster version of at least 1.7 to support initializers:
```
Comment (ccarpentiere) [inline tutorials/transparent-proxy-and-filtering-on-k8s-with-initializers/index.md]: "Create a Container Engine cluster with alpha features enabled, RBAC support, and..."

File: tutorials/transparent-proxy-and-filtering-on-k8s-with-initializers/index.md [hunk 7]
```diff
@@ -0,0 +1,216 @@
...
+
+        gcloud container clusters create tproxy-example \
+        --zone us-central1-f \
+        --machine-type n1-standard-1 \
+        --num-nodes 3 \
+        --enable-kubernetes-alpha \
+        --cluster-version 1.7.6 \
+        --no-enable-legacy-authorization
+
+    This command will also automatically configure the kubectl command to use the cluster.
```
Comment (ccarpentiere) [inline tutorials/transparent-proxy-and-filtering-on-k8s-with-initializers/index.md]: Use present tense here & throughout, "This command also automatically configures the kubectl command to use the cluster."

File: tutorials/transparent-proxy-and-filtering-on-k8s-with-initializers/index.md [hunk 8]
```diff
@@ -0,0 +1,216 @@
...
+
+2. Create service account and cluster role binding for Helm to enable RBAC support:
```
Comment (ccarpentiere) [inline tutorials/transparent-proxy-and-filtering-on-k8s-with-initializers/index.md]: "Create a service account..."

File: tutorials/transparent-proxy-and-filtering-on-k8s-with-initializers/index.md [hunk 9]
```diff
@@ -0,0 +1,216 @@
...
+
+3. Install the Helm tool locally in your Cloud Shell instance:
+
+        curl -sL https://storage.googleapis.com/kubernetes-helm/helm-v2.5.1-linux-amd64.tar.gz | tar -xvf - && sudo mv linux-amd64/helm /usr/local/bin/ && rm -Rf linux-amd64
+
+4. Initialize Helm with the service account:
+
+        helm init --service-account=tiller
+
+    This will install the server side component of Helm, Tiller,  in the Kubernetes cluster. The Tiller pod may take a minute to start, run the command below to verify it has been deployed:
```
Comment (ccarpentiere) [inline tutorials/transparent-proxy-and-filtering-on-k8s-with-initializers/index.md]: "This installs.."

File: tutorials/transparent-proxy-and-filtering-on-k8s-with-initializers/index.md [hunk 10]
```diff
@@ -0,0 +1,216 @@
...
+        helm version
+
+    You should see the Client and Server versions in the output:
+
+        Client: &version.Version{SemVer:"v2.5.1", GitCommit:"7cf31e8d9a026287041bae077b09165be247ae66", GitTreeState:"clean"}
+        Server: &version.Version{SemVer:"v2.5.1", GitCommit:"7cf31e8d9a026287041bae077b09165be247ae66", GitTreeState:"clean"}
+
+## Install the Helm chart
+
+Before installing the chart, you must first extract the certificates generated by mitmproxy. The generated CA cert will be used in the example pods to trust the proxy when making HTTPS requests.
```
Comment (ccarpentiere) [inline tutorials/transparent-proxy-and-filtering-on-k8s-with-initializers/index.md]: "The generated CA cert is used..."

**Comment by danisla** [comment]
@ccarpentiere, changes incorporated.

File: tutorials/transparent-proxy-and-filtering-on-k8s-with-initializers/index.md [hunk 11]
```diff
@@ -0,0 +1,214 @@
...
+
+This tutorial assumes you already have a Google Cloud Platform (GCP) account and are familiar with the high level concepts of Kubernetes Pods and Deployments.
+
+## Costs
+
+This tutorial uses billable components of GCP, including:
+
+- [Google Container Engine](https://cloud.google.com/container-engine/pricing)
+
+Use the [Pricing Calculator](https://cloud.google.com/products/calculator/#id=f52c2651-4b02-4da3-b8cd-fdbca6ad89a9) to estimate the costs for your environment. The resources used in this tutorial are approximately $2.39 per day.
```
Comment (ahmetb) [inline tutorials/transparent-proxy-and-filtering-on-k8s-with-initializers/index.md]: don't include the price as it can easily change.

File: tutorials/transparent-proxy-and-filtering-on-k8s-with-initializers/index.md [hunk 12]
```diff
@@ -0,0 +1,214 @@
...
+        cd kubernetes-tproxy
+
+    The remainder of this tutorial will be run from the root of the cloned repository directory.
+
+## Create Container Engine cluster and install Helm
+
+1. Create Container Engine cluster with alpha features enabled, RBAC support, and a cluster version of at least 1.7 to support initializers:
+
+        gcloud container clusters create tproxy-example \
+        --zone us-central1-f \
```
Comment (ahmetb) [inline tutorials/transparent-proxy-and-filtering-on-k8s-with-initializers/index.md]: indent lines by 2 spaces after the first line

File: tutorials/transparent-proxy-and-filtering-on-k8s-with-initializers/index.md [hunk 13]
```diff
@@ -0,0 +1,214 @@
...
+
+        kubectl create serviceaccount tiller --namespace kube-system
+
+        kubectl create clusterrolebinding tiller-cluster-rule \
+        --clusterrole=cluster-admin \
+        --serviceaccount=kube-system:tiller
+
+3. Install the Helm tool locally in your Cloud Shell instance:
+
+        curl -sL https://storage.googleapis.com/kubernetes-helm/helm-v2.5.1-linux-amd64.tar.gz | tar -xvf - && sudo mv linux-amd64/helm /usr/local/bin/ && rm -Rf linux-amd64
```
Comment (ahmetb) [inline tutorials/transparent-proxy-and-filtering-on-k8s-with-initializers/index.md]: can you refer to helm docs here? you're hardcoding multiple things that can easily stop working tomorrow.

File: tutorials/transparent-proxy-and-filtering-on-k8s-with-initializers/index.md [hunk 14]
```diff
@@ -0,0 +1,214 @@
...
+
+1. Create Container Engine cluster with alpha features enabled, RBAC support, and a cluster version of at least 1.7 to support initializers:
+
+        gcloud container clusters create tproxy-example \
+        --zone us-central1-f \
+        --machine-type n1-standard-1 \
+        --num-nodes 3 \
+        --enable-kubernetes-alpha \
+        --cluster-version 1.7.6 \
+        --no-enable-legacy-authorization
```
Comment (ahmetb) [inline tutorials/transparent-proxy-and-filtering-on-k8s-with-initializers/index.md]: why is this necessary?

File: tutorials/transparent-proxy-and-filtering-on-k8s-with-initializers/index.md [hunk 15]
```diff
@@ -0,0 +1,214 @@
...
+## Create Container Engine cluster and install Helm
+
+1. Create Container Engine cluster with alpha features enabled, RBAC support, and a cluster version of at least 1.7 to support initializers:
+
+        gcloud container clusters create tproxy-example \
+        --zone us-central1-f \
+        --machine-type n1-standard-1 \
+        --num-nodes 3 \
+        --enable-kubernetes-alpha \
+        --cluster-version 1.7.6 \
```
Comment (ahmetb) [inline tutorials/transparent-proxy-and-filtering-on-k8s-with-initializers/index.md]: why do you need to specify cluster-version?

File: tutorials/transparent-proxy-and-filtering-on-k8s-with-initializers/index.md [hunk 16]
```diff
@@ -0,0 +1,214 @@
...
+
+## Create Container Engine cluster and install Helm
+
+1. Create Container Engine cluster with alpha features enabled, RBAC support, and a cluster version of at least 1.7 to support initializers:
+
+        gcloud container clusters create tproxy-example \
+        --zone us-central1-f \
+        --machine-type n1-standard-1 \
+        --num-nodes 3 \
+        --enable-kubernetes-alpha \
```
Comment (ahmetb) [inline tutorials/transparent-proxy-and-filtering-on-k8s-with-initializers/index.md]: make sure to come back & update the version flag and this flag after 1.9.x is out on GKE

File: tutorials/transparent-proxy-and-filtering-on-k8s-with-initializers/index.md [hunk 17]
```diff
@@ -0,0 +1,214 @@
...
+
+        Add this metadata annotation to your deployment specs to apply the tproxy initializer:
+
+        metadata:
+        annotations:
+            "initializer.kubernetes.io/tproxy": "true"
+
+3. Run the command below to get the status of the DaemonSet pods:
+
+        kubectl get pods -o wide
```
Comment (ahmetb) [inline tutorials/transparent-proxy-and-filtering-on-k8s-with-initializers/index.md]: Looks like you got an extra space at the beginning of the command according to GitHub markdown rendering. This will cause bash to not to remember the command.

File: tutorials/transparent-proxy-and-filtering-on-k8s-with-initializers/index.md [hunk 18]
```diff
@@ -0,0 +1,214 @@
...
+title: Transparent Proxy and Filtering on Kubernetes with Initializers
+description: Learn how to run a transparent proxy on Kubernetes to filter and intercept traffic out of your deployments using custom deployment initializers.
+author: danisla
+tags: Kubernetes mitmproxy proxy
+date_published: 2017-09-23
+---
+
+Dan Isla | Google Cloud Solution Architect | Google
+
+This is a follow-on tutorial to the [Transparent Proxy and Filtering on K8S](https://cloud.google.com/community/tutorials/transparent-proxy-and-filtering-on-k8s) tutorial. It shows how to simplify the application of a transparent proxy for existing deployments using a Deployment Initializer. Initializers are one of the [Dynamic Admission Control](https://kubernetes.io/docs/admin/extensible-admission-controllers/) features of Kubernetes, and are available as an alpha feature in Kubernetes 1.7.
```
Comment (ahmetb) [inline tutorials/transparent-proxy-and-filtering-on-k8s-with-initializers/index.md]: K8S &rarr; Kubernetes

**Comment by jmdobry** [comment]
I'll do my review when @ahmetb's comments are addressed.

File: tutorials/transparent-proxy-and-filtering-on-k8s-with-initializers/index.md (hunk 11 above)
Comment (danisla) [reply tutorials/transparent-proxy-and-filtering-on-k8s-with-initializers/index.md]: I agree that it can change but I'll defer to @ccarpentiere or @jmdobry on what we should do here because I've had tech writers ask me to add the price when it wasn't there.

File: tutorials/transparent-proxy-and-filtering-on-k8s-with-initializers/index.md (hunk 13 above)
Comment (danisla) [reply tutorials/transparent-proxy-and-filtering-on-k8s-with-initializers/index.md]: Helm 2.6 has an issue with the initializer object creation. I want to hard code it. The tutorial (any my other tutorials) are under CI to catch breaking changes.

File: tutorials/transparent-proxy-and-filtering-on-k8s-with-initializers/index.md (hunk 17 above)
Comment (danisla) [reply tutorials/transparent-proxy-and-filtering-on-k8s-with-initializers/index.md]: Hmm, I don't see it, can you confirm there is an issue?

File: tutorials/transparent-proxy-and-filtering-on-k8s-with-initializers/index.md (hunk 16 above)
Comment (danisla) [reply tutorials/transparent-proxy-and-filtering-on-k8s-with-initializers/index.md]: yup, will do.

File: tutorials/transparent-proxy-and-filtering-on-k8s-with-initializers/index.md (hunk 17 above)
Comment (ahmetb) [reply tutorials/transparent-proxy-and-filtering-on-k8s-with-initializers/index.md]: I can't confirm as I'm not capable of doing what devsite markdown renderer (or whatever we use) does. If I click preview for this file, I see a trailing extra space on GitHub UI that's all I wanted to say.

File: tutorials/transparent-proxy-and-filtering-on-k8s-with-initializers/index.md (hunk 18 above)
Comment (danisla) [reply tutorials/transparent-proxy-and-filtering-on-k8s-with-initializers/index.md]: also need to match the URL with the deployed link of the other tutorial, @jmdobry do you know what that will be?

File: tutorials/transparent-proxy-and-filtering-on-k8s-with-initializers/index.md (hunk 14 above)
Comment (danisla) [reply tutorials/transparent-proxy-and-filtering-on-k8s-with-initializers/index.md]: yeah, the only required arg now is `--enable-kubernetes-alpha`, I'll remove the others.

File: tutorials/transparent-proxy-and-filtering-on-k8s-with-initializers/index.md (hunk 17 above)
Comment (danisla) [reply tutorials/transparent-proxy-and-filtering-on-k8s-with-initializers/index.md]: I think it will be ok, my other tutorials are written this way and the final conversion to the website looks OK.

File: tutorials/transparent-proxy-and-filtering-on-k8s-with-initializers/index.md (hunk 18 above)
Comment (jmdobry) [reply tutorials/transparent-proxy-and-filtering-on-k8s-with-initializers/index.md]: #251 will end up at https://cloud.google.com/community/tutorials/transparent-proxy-and-filtering-on-k8s

#252 will end up at https://cloud.google.com/community/tutorials/transparent-proxy-and-filtering-on-k8s-with-initializers

Comment (danisla) [reply tutorials/transparent-proxy-and-filtering-on-k8s-with-initializers/index.md]: Great, thanks!

**Comment by jmdobry** [comment]
@ahmetb Can you approve the PR?

File: tutorials/transparent-proxy-and-filtering-on-k8s-with-initializers/index.md (hunk 18 above)
Comment (jmdobry) [reply tutorials/transparent-proxy-and-filtering-on-k8s-with-initializers/index.md]: I fixed the K8S -> Kubernetes

**Comment by jmdobry** [comment]
Awaiting final approval, and blocked on #251

File: tutorials/transparent-proxy-and-filtering-on-k8s-with-initializers/index.md (hunk 11 above)
Comment (jmdobry) [reply tutorials/transparent-proxy-and-filtering-on-k8s-with-initializers/index.md]: The link to the calculator is sufficient

Comment (danisla) [reply tutorials/transparent-proxy-and-filtering-on-k8s-with-initializers/index.md]: Ok. I remove the price and left the link.

**Comment by jmdobry** [comment]
Published: https://cloud.google.com/community/tutorials/transparent-proxy-and-filtering-on-k8s-with-initializers
